### PR TITLE
Lazy load type graph

### PR DESCRIPTION
**Pull request by siyuanh**

### CODE REVIEW

**Comment by siyuanh** [comment]
Jenkins test this please

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/FastClassIndexReader.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+package com.datatorrent.stram.webapp.asm;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+
+import org.apache.xbean.asm5.Opcodes;
+
+/**
+ * Created by siyuan on 10/29/15.
```
Comment (otterc) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/FastClassIndexReader.java]: Author name in source file?

**Comment by otterc** [comment]
Are the line lengths less than 120? Since you are fixing existing style in these classes maybe you should do this as well

File: engine/src/test/java/com/datatorrent/stram/webapp/asm/FastClassIndexReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.xbean.asm5.ClassReader;
+
+/**
+ * Created by siyuan on 10/30/15.
```
Comment (otterc) [inline engine/src/test/java/com/datatorrent/stram/webapp/asm/FastClassIndexReaderTest.java]: author name ?

File: engine/src/main/java/com/datatorrent/stram/webapp/asm/FastClassIndexReader.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+  {
+    for (int l = 0; l < subArray.length; l++) {
+      if (bb[i + l] != subArray[l]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public int readUnsignedShort(final int index)
```
Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/asm/FastClassIndexReader.java]: public function but has no documentation

File: engine/src/test/java/com/datatorrent/stram/webapp/OperatorDiscoveryTest.java [hunk 4]
```diff
@@ -1138,5 +1140,7 @@ public void testMethodType()
     Assert.assertEquals("@description", OperatorDiscoverer.MethodTagType.DESCRIPTION, OperatorDiscoverer.MethodTagType.from("@description"));
     Assert.assertEquals("@random", null, OperatorDiscoverer.MethodTagType.from("@random"));
   }
+
```
Comment (gauravgopi123) [inline engine/src/test/java/com/datatorrent/stram/webapp/OperatorDiscoveryTest.java:1141]: extra line

File: engine/src/test/java/com/datatorrent/stram/webapp/OperatorDiscoveryTest.java [hunk 5]
```diff
@@ -24,6 +24,7 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
+import java.io.IOException;
```
Comment (ishark) [inline engine/src/test/java/com/datatorrent/stram/webapp/OperatorDiscoveryTest.java]: No new change in this file.. Unused imports?

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 6]
```diff
@@ -602,6 +558,14 @@ public void setJarName(String jarName)
 
     public CompactClassNode getClassNode()
     {
+      if (classNode == null) {
```
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java:566]: Do we need double-checked locking here? Very slim chance, but is it likely that describeOperator  gets called simultaneously through dtcli?

Comment (siyuanh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java:566]: Good catch, this method should be synchronized

**Comment by siyuanh** [comment]
@ishark  I have updated the pull request, please take a look

**Pull request merged**